tree/treeplayer/src/TTreeFormula.cxx

```diff
@@ -3412,8 +3412,24 @@ Int_t TTreeFormula::GetRealInstance(Int_t instance, Int_t codeindex) {
                }
                break;
             case -1: {
-                  local_index = 0;
-                  Int_t virt_accum = 0;
+	      // caches to keep track of where we are across function invocations
+	      // will avoid an O(N^2) scaling
+	      // reset them if instance=0 is seen
+              static Int_t local_index_cache = 0;
```

Comment:
actually I am very wrong :( ...

Those actually need to be data member.   More than once TTreeFormula can be handled/processed during the same operation and they can also be nested.  For example
```
tree->Draw("px:pz:fTrack[index]");
```
the call to EvalInstance for each i will be interleaved and the one with the index will be nested.